src/main/scala/tech/minna/utilities/TraitEnumeration.scala

```diff
@@ -0,0 +1,34 @@
+package tech.minna.utilities
+
+import scala.language.experimental.macros
+import scala.reflect.macros.blackbox
+import scala.reflect.internal.Symbols
+
+object TraitEnumeration {
+  def values[A]: Set[A] = macro valuesImpl[A]
+
+  def valuesImpl[A: c.WeakTypeTag](c: blackbox.Context) = {
+    import c.universe._
+
+    val symbol = weakTypeOf[A].typeSymbol.asClass
+
+    if (!symbol.isClass || !symbol.isSealed)
+      c.abort(c.enclosingPosition, "Can only enumerate values of a sealed trait or class.")
```

Comment:
"Can only generate enumeration values of a sealed trait that is only extended by case objects."